### ISSUE TITLE
Add PCONTEXT/CONTEXT in ffi.rs for ARM

### ISSUE THREAD

**Issue by chandde** [alexcrichton/backtrace-rs#180]
I've added the same for winapi-rs, see https://github.com/chandde/winapi-rs/blob/0.3/src/um/winnt.rs#L1482-L1525, and I have PR to upstream the change. Seems I need to do similar things for backtrace in ffi.rs, however I'm not exactly sure how I'm gonna define the union inside CONTEXT

```
STRUCT!{struct NEON128 {
    Low: ULONGLONG,
    High: LONGLONG,
}}
pub type PNEON128 = *mut NEON128;
UNION!{union CONTEXT_u {
    [u64; 32],
    Q Q_mut: [NEON128; 16],
    D D_mut: [ULONGLONG; 32],
    S S_mut: [DWORD; 32],
}}
```
I noticed the other CONTEXT for AMD64, which also has a union
```
UNION!{union CONTEXT_u {
    [u64; 64],
    FltSave FltSave_mut: XMM_SAVE_AREA32,
    s s_mut: CONTEXT_u_s,
}}
```
it's used in the struct as a single line     `u: CONTEXT_u,`, however I see you put FltSave directly into the struct as `pub FltSave: FLOATING_SAVE_AREA`, and ignored `FltSave` in the macro rules for ffi
```
    (@test_fields $a:ident $b:ident FltSave $ty:ty) => (
        // Skip this field on x86_64 `CONTEXT` since it's a union and a bit funny
    );
```
So any suggestions how I should put the 3 fields (Q, D, S) union inside the CONTEXT struct?

Thanks

**Comment by alexcrichton** [comment]
Ah yeah sorry I was having some difficulty here as well mapping things out. I'd take a look at x86 rather than aarch64 which uses the [`D` field](https://github.com/alexcrichton/backtrace-rs/blob/8018d458ad26a1fca3453ef155d3c1ff07f52887/src/dbghelp/ffi.rs#L90-L95) of a [union below](https://github.com/alexcrichton/backtrace-rs/blob/8018d458ad26a1fca3453ef155d3c1ff07f52887/src/dbghelp/ffi.rs#L400-L403).

Basically just put one member of the union into a struct and then edit the verification macro to call a method on winapi but access a field for this crate. If that still causes difficulty I can help out!

**Comment by chandde** [comment]
OK, here's what I have now, can you take a look? https://github.com/alexcrichton/backtrace-rs/compare/master...chandde:master

I cannot send PR until my other PR for winapi is merged and its version is bumped, however, locally I patched backtrace using my fork, and it built successfully. Although I don't understand why the code I added threw a warning about using unsafe, but existing code did not. Scroll to the bottom to see the warning.

```
D:\git\backtrace-rs>cargo update -p winapi --precise 9b9f3dd
    Updating git repository `https://github.com/chandde/winapi-rs`
    Updating crates.io index
      Adding winapi v0.3.7 (https://github.com/chandde/winapi-rs?branch=0.3#9b9f3dde)
    Removing winapi v0.3.7 (https://github.com/chandde/winapi-rs?branch=prtest3#b4400bd5)
      Adding winapi-i686-pc-windows-gnu v0.4.0 (https://github.com/chandde/winapi-rs?branch=0.3#9b9f3dde)
    Removing winapi-i686-pc-windows-gnu v0.4.0 (https://github.com/chandde/winapi-rs?branch=prtest3#b4400bd5)
      Adding winapi-x86_64-pc-windows-gnu v0.4.0 (https://github.com/chandde/winapi-rs?branch=0.3#9b9f3dde)
    Removing winapi-x86_64-pc-windows-gnu v0.4.0 (https://github.com/chandde/winapi-rs?branch=prtest3#b4400bd5)

D:\git\backtrace-rs>findstr winapi Cargo.*
Cargo.lock: "winapi 0.3.7 (git+https://github.com/chandde/winapi-rs?branch=0.3)",
Cargo.lock: "winapi 0.3.7 (git+https://github.com/chandde/winapi-rs?branch=0.3)",
Cargo.lock: "winapi 0.3.7 (git+https://github.com/chandde/winapi-rs?branch=0.3)",
Cargo.lock: "winapi 0.3.7 (git+https://github.com/chandde/winapi-rs?branch=0.3)",
Cargo.lock: "winapi 0.3.7 (git+https://github.com/chandde/winapi-rs?branch=0.3)",
Cargo.lock:name = "winapi"
Cargo.lock:source = "git+https://github.com/chandde/winapi-rs?branch=0.3#9b9f3ddec4079d01f0b8de580066ca2166661259"
Cargo.lock: "winapi-i686-pc-windows-gnu 0.4.0 (git+https://github.com/chandde/winapi-rs?branch=0.3)",
Cargo.lock: "winapi-x86_64-pc-windows-gnu 0.4.0 (git+https://github.com/chandde/winapi-rs?branch=0.3)",
Cargo.lock:name = "winapi-i686-pc-windows-gnu"
Cargo.lock:source = "git+https://github.com/chandde/winapi-rs?branch=0.3#9b9f3ddec4079d01f0b8de580066ca2166661259"
Cargo.lock:name = "winapi-x86_64-pc-windows-gnu"
Cargo.lock:source = "git+https://github.com/chandde/winapi-rs?branch=0.3#9b9f3ddec4079d01f0b8de580066ca2166661259"
Cargo.lock:"checksum winapi 0.3.7 (git+https://github.com/chandde/winapi-rs?branch=0.3)" = "<none>"
Cargo.lock:"checksum winapi-i686-pc-windows-gnu 0.4.0 (git+https://github.com/chandde/winapi-rs?branch=0.3)" = "<none>"
Cargo.lock:"checksum winapi-x86_64-pc-windows-gnu 0.4.0 (git+https://github.com/chandde/winapi-rs?branch=0.3)" = "<none>"
Cargo.toml:winapi = { version = "^0.3.7", optional = true }
Cargo.toml:verify-winapi = [
Cargo.toml:  'winapi/dbghelp',
Cargo.toml:  'winapi/processthreadsapi',
Cargo.toml:  'winapi/winnt',
Cargo.toml:  'winapi/minwindef',
Cargo.toml:  'winapi/libloaderapi',
Cargo.toml:winapi = { git = "https://github.com/chandde/winapi-rs", branch = "0.3" }
D:\git\backtrace-rs>cargo clean

D:\git\backtrace-rs>cargo build --release --all --target thumbv7a-pc-windows-msvc --all-features
   Compiling proc-macro2 v0.4.30
   Compiling semver-parser v0.7.0
   Compiling unicode-xid v0.1.0
   Compiling syn v0.15.34
   Compiling arrayvec v0.4.10
   Compiling cfg-if v0.1.7
   Compiling glob v0.2.11
   Compiling byteorder v1.3.1
   Compiling libc v0.2.54
   Compiling winapi v0.3.7 (https://github.com/chandde/winapi-rs?branch=0.3#9b9f3dde)
   Compiling plain v0.2.3
   Compiling cc v1.0.36
   Compiling nodrop v0.1.13
   Compiling fallible-iterator v0.2.0
   Compiling autocfg v0.1.2
   Compiling serde v1.0.91
   Compiling stable_deref_trait v1.1.1
   Compiling uuid v0.7.4
   Compiling smallvec v0.6.9
   Compiling lazy_static v1.3.0
   Compiling lazycell v1.2.1
   Compiling rustc-demangle v0.1.14
   Compiling rustc-serialize v0.3.24
   Compiling log v0.4.6
   Compiling findshlibs v0.4.1
   Compiling semver v0.9.0
   Compiling backtrace v0.3.15 (D:\git\backtrace-rs)
   Compiling cpp_demangle v0.2.12
   Compiling intervaltree v0.2.4
   Compiling gimli v0.18.0
   Compiling rustc_version v0.2.3
   Compiling scroll v0.9.2
   Compiling backtrace-sys v0.1.28 (D:\git\backtrace-rs\backtrace-sys)
   Compiling quote v0.6.12
   Compiling memmap v0.7.0
   Compiling scroll_derive v0.9.5
   Compiling serde_derive v1.0.91
   Compiling goblin v0.0.22
   Compiling object v0.12.0
   Compiling addr2line v0.9.0
warning: unnecessary `unsafe` block
   --> src\backtrace\dbghelp.rs:191:5
    |
191 |     unsafe {
    |     ^^^^^^ unnecessary `unsafe` block
    |
    = note: #[warn(unused_unsafe)] on by default

    Finished release [optimized] target(s) in 1m 02s

D:\git\backtrace-rs>
```

**Comment by alexcrichton** [comment]
That looks good to me! I think you can probably just remove that `u nsafe` block and it shoudl be good to go

**Comment by chandde** [comment]
OK, I'll close this issue and send a PR right away, even the winapi PR is not merged yet. The truth is I'm probably the only one that's compiling rust for ARM Windows, and my changes to backtrace is guarded by "arm", so it should have no impact to the rest. When the winapi PR is merged and a new version is released, once we switch to the public winapi, we'll run into issues in backtrace and that'll be the time we revisit backtrace and bump the dependency version of winapi in backtrace toml.
